improver/calibration/reliability_calibration.py

```diff
@@ -0,0 +1,388 @@
+        """
+        Extract a truth field that matches the threshold and time of the
+        forecast field.
+
+        Args:
+            truths (iris.cube.Cube):
+                The cube containing the thresholded truths used in calibration.
+            forecast_slice (iris.cube.Cube):
+                A slice over threshold and time of the forecast data.
+            threshold_coord (iris.coords.DimCoord):
+                The threshold coordinate that should be present on both input
+                cubes.
+        Returns:
+            iris.cube.Cube or None:
+                A truth cube that matches the time and threshold values of the
+                forecast slice, or None if no match is found.
+        """
+        threshold, = forecast_slice.coord(threshold_coord).points
+        threshold_name = threshold_coord.name()
+        threshold_constraint = {threshold_name: threshold}
+        time = forecast_slice.coord('time')
+
+        return truths.extract(iris.Constraint(
+            time=next(time.cells()).point, coord_values=threshold_constraint))
+
+    def process(self, historic_forecasts, truths):
+        """
+        Slice data over threshold and time coordinates to construct reliability
+        tables. These are summed over time to give a single table for each
+        threshold, constructed from all the provided historic forecasts and
+        truths.
+
+        Args:
+            historic_forecasts (iris.cube.Cube):
+                A cube containing the historical forecasts used in calibration.
+            truths (iris.cube.Cube):
+                A cube containing the thresholded gridded truths used in
+                calibration.
+        Returns:
+            iris.cube.CubeList:
+                A cubelist of reliability table cubes, one for each threshold
+                in the historic forecast cubes.
+        """
+        threshold_coord = find_threshold_coordinate(historic_forecasts)
+        time_coord = historic_forecasts.coord('time')
+
+        reliability_tables = iris.cube.CubeList()
+        for threshold_slice in historic_forecasts.slices_over(threshold_coord):
+
+            threshold_reliability = []
+            for time_slice in threshold_slice.slices_over(time_coord):
+
+                truth = self._extract_truth(truths, time_slice,
+                                            threshold_coord)
+                if not truth:
+                    continue
+
+                reliability_table = (
+                    self._populate_reliability_bins(
+                        time_slice.data, truth.data))
+
+                threshold_reliability.append(reliability_table)
+
+            # Stack and sum reliability tables for all times
+            table_values = np.stack(threshold_reliability)
+            table_values = np.sum(table_values, axis=0, dtype=np.float32)
+
+            reliability_cube = self._create_reliability_table_cube(
```

Comment:
I was wondering about the idea of potentially moving the call to `_create_reliability_table_cube` to prior to any of the `for` loops in this `process` method. As far as I can tell most of the work that `_create_reliability_table_cube` does, could be done up front to create a template cube e.g. `reliability_cube_template = self._create_reliability_table_cube()`. This line inside the `for` loop would then be replaced with `reliability_cube = reliability_cube_template.copy(data=table_values)`. This would avoid having to recreate the reliability_table_cube in each iteration, and also avoid the Codacy issue related to the undefined loop variable 'time_slice'. This would also mean that the `cycle_hour` issue, related to a possible mismatch in cycle_hour going unnoticed, could be dealt with prior to the `for` loops.

Reply:
Done.